source/nodejs/adaptivecards/src/serialization.ts

```diff
@@ -111,6 +116,22 @@ export class Versions {
     static readonly v1_5 = new Version(1, 5);
     static readonly v1_6 = new Version(1, 6);
     static readonly latest = Versions.v1_6;
+    // version over latest is considered "vNext feature"
+    // if latest or maxSupportedVersion is changed,
+    // don't forget to update styles .ac-schema-version-1-?::after to add/remove Preview tag from adaptivecards-site\themes\adaptivecards\source\css\style.css
+    static readonly maxSupportedVersion = Versions.v1_6;
```

Comment:
Why is this necessary when `Versions.latest` already means the same?

Reply:
`latest` is a little confusion here. actually it means 'stable'. maybe I need to rename it to stableVersion
 the version above `stable` but <= maxSupportedVresion will be the preview features. 
If it's above maxSupportedVersion, it raises `The specified card version (${version}) is not supported`, and that's why maxSupportedVersion is introduced.

Reply:
We can't rename `Versions.latest` as it's a shipped API and changing it would risk breaking application code when it's not really necessary.

For the rest, please help me understand. Currently you have this:

```typescript
export class Versions {
    static readonly v1_0 = new Version(1, 0);
    static readonly v1_1 = new Version(1, 1);
    static readonly v1_2 = new Version(1, 2);
    static readonly v1_3 = new Version(1, 3);
    static readonly v1_4 = new Version(1, 4);
    static readonly v1_5 = new Version(1, 5);
    static readonly v1_6 = new Version(1, 6);
    static readonly latest = Versions.v1_5;
    static readonly maxSupportedVersion = Versions.v1_6;
}
```

You're saying that if `Versions.latest < version <= Versions.maxSupportedVersion` then `version` is a preview version. How is that different from just saying `version` is a preview if `version > Versions.latest`? When would you ever hit a situation where there is another declared version that is > maxSupportedVersion but isn't a preview version?

I'm fairly certain `maxSupportedVersion` isn't necessary.

Furthermore, if you want to be granular in terms of which version is a preview version, you are probably better off making that an explicit flag in the version instance itself:

```typescript
export class Versions {
    static readonly v1_0 = new Version(1, 0);
    static readonly v1_1 = new Version(1, 1);
    static readonly v1_2 = new Version(1, 2);
    static readonly v1_3 = new Version(1, 3);
    static readonly v1_4 = new Version(1, 4);
    static readonly v1_5 = new Version(1, 5);
    static readonly v1_6 = new Version(1, 6, undefined /* label */, true /* isPreview */);
    static readonly latest = Versions.v1_5;
}
```

The above would make v1_6 and v1_6 only a preview version.

Reply:
I was thinking to add isPreview parameter to Version as you mentioned. but I finally dropped it.

Here is the logic:
if (version <= latest) {
   Ok
} else if (version <= maxVersion) {
   raise "it's preview"
} else {
   raise "The specified card version (${version}) is not supported"
}

If I implements 'isPreview' in version, the logic 
```
if version in [v1_0, v1_1, v1_2.... v1_7] {
      if (version.isPreview() {}
     else {
    }
} else {
}
```

So 
1. I need to implement new logic to convert "1.7' to a valid Version first or implement a new comparator
2. make v1_6, v1_7 somewhere too to do the comparison. Then if I add a new version, I will need to make change on multiple place.

In these solution, I only need to change maxSupportedVersion, and change are grouped together.

Reply:
Sorry, I still don't get it. The goal seems to be to warn the user in the designer that they're using a property value that's only supported in a version that happens to be a preview version. So all you need to know is that a version is a preview, so a flag on the Version class would work just fine.

Besides, the Version class already has a `label` property that is designed to and has been used for this type of thing. Essentially, in your case, you'd want to define v1_6 as follows:
```typescript
export class Versions {
    static readonly v1_6 = new Version(1, 6, "1.6 (Preview)");
}
```

That's enough for the version picker in the designer to show "1.6 (Preview)" and thus let the user know they're authoring their card against a preview version.

Now the designer already has the property warning you seem to want:
![image](https://user-images.githubusercontent.com/1334689/142298315-05a20734-9328-43d7-884f-3ba484b9ad88.png)

But, it so happens that it is emitted with this code:
![image](https://user-images.githubusercontent.com/1334689/142298393-f8f0a4c2-ef4d-43f7-b661-85cbc054ce02.png)

Because it uses `toString()`, the label isn't displayed. If you just changed that to use `targetVersion.label` instead, you'd get exactly what you're after without needing to introduce anything new.

Reply:
In the end, using `Version.label` will yield this result:
![image](https://user-images.githubusercontent.com/1334689/142299720-9502a5cc-dbf9-4d7b-893c-cc9849fef881.png)

Which is basically what you want based on the screenshot in the description of this PR.

Reply:
I see code like this:

```
    static parse(versionString: string, context: BaseSerializationContext): Version | undefined {
        if (!versionString) {
            return undefined;
        }

        let result = new Version();
        result._versionString = versionString;
```

How do you check if the parsed version is preview or not?

Reply:
Why would I need to if all I need is a designer warning?

But if you _do_ need it, then an `isPreview` property on `Version` is the way to go.

Reply:
```typescript
export class Version {
    ...
    private _isPreview: boolean;
    ...
    constructor(major: number = 1, minor: number = 1, label?: string, isPreview: boolean = false) {
        ...
        this._isPreview = isPreview;
    }

    get isPreview(): boolean {
        return this._isPreview;
    }
}

```

Reply:
Oh, and of course you'd need to:
- Update the `Version.parse` logic...
- ...and thus introduce a version format that makes it possible to identify the version string as a preview

But I really don't see why you'd need all that. Using the label should be all you need to achieve the result you want, unless I am missing one of your requirements.